modules/ding_wayf_dk/includes/DingWayfGateway.class.inc

```diff
@@ -0,0 +1,118 @@
+<?php
+/**
+ * @file
+ * Simple class for communication with the SimpleSAMLphp DBC gateway.
+ */
+
+namespace WAYF;
+use WAYF\SPortoException;
+
+/**
+ * DingWayfGateway class
+ */
+class DingWayfGateway implements WayfServiceInterface {
+  protected $config = array();
+
+  /**
+   * Default construct for the class.
+   *
+   * @param array $config
+   *   WAYF configuration. This is not used by the object but it's the default
+   *   parameter when the class is created in the wayf_dk module. It's used by
+   *   other service classes. So here it's simply overridden with gateway WAYF
+   *   settings.
+   */
+  public function __construct(array $config) {
+    $this->config = $config;
+
+    // Note: This is a config override hack.
+    $this->config['gateway'] = variable_get('ding_wayf_dk_gateway_endpoint', '');
+    $this->config['return'] = array(
+      'login' => variable_get('ding_wayf_dk_gateway_consumer', ''),
+      'logout' => variable_get('ding_wayf_dk_gateway_logout_endpoint', ''),
+    );
+  }
+
+  /**
+   * Authentication with the Single SignOn Service.
+   *
+   * @param array $providerids
+   *   Provider ids are used for scoping in the wayf_dk_login service class, but
+   *   it's not used by this service class.
+   *
+   * @return array
+   *   The response from the gateway.
+   *
+   * @throws \WAYF\SPortoException
+   *   Is thrown on gateway error.
+   */
+  public function authenticate($providerids = array()) {
+    // Check error callback.
+    $error = isset($_GET['error']) ? $_GET['error'] : FALSE;
+    if ($error) {
+      watchdog('ding_wayf', 'The gateway returned an error: %error', array('%error' => $error), WATCHDOG_ERROR);
+      throw new SPortoException('The gateway returned an error: ' . $error);
+    }
+
+    // The gateway returns attributes in the $_POST array, if authentication
+    // have gone well.
+    if (empty($_POST['eduPersonTargetedID'])) {
+
+      // Redirect the user by setting the location header.
+      header('Location:' . $this->config['gateway'] . '?returnUrl=' . $this->config['return']['login'] . '&idp=wayf');
+      exit;
+    }
+    else {
+      if (!empty($_POST['eduPersonTargetedID'])) {
```

Comment:
This `if` statement does not make sense. We already check whether `$_POST['eduPersonTargetedID']` is empty on L59.